cpp/streamer/impl/batch/batch.cc

```diff
@@ -139,24 +157,34 @@ void Batch::read(const Config & config)
         finished_until(file_offset, common::ResponseCode::Success);
     }
 
-    if (file_offset < range.end)
+    if (file_offset < range.end && !stopped)
     {
         _reader->read(range.end - file_offset, buffer);
         finished_until(range.end, common::ResponseCode::Success);
     }
 
-    LOG(DEBUG) << "Finished reading successfuly from file " << path;
+    LOG(DEBUG) << "Finished reading " << i << "/" << num_chunks << " chunks of size " << config.fs_block_bytesize << " from file " << path << (stopped ? " - terminated" : " successfully");
+
+    if (stopped)
+    {
+        throw common::Exception(common::ResponseCode::FinishedError);
+    }
 }
 
 // read the entire range and send notifications for each sub range
-void Batch::async_read(const Config & config)
+void Batch::async_read(const Config & config, std::atomic<bool> & stopped)
 {
     if (tasks.empty())
     {
         LOG(DEBUG) << "Empty batch";
         return;
     }
 
+    if (stopped)
```

Comment:
What is the purpose of this check?
These line of code are called right after the execute called.
Its just that I dont see you aquire a lock before.
And you right after sends the requests. Why not sending the requests and then if it was cacnceled immediately the async response will return finished_error?

Reply:
The purpose is to avoid sending requests if the streamer is already stopped. The stopped flag is not checked in `execute`. It is the same as in the synchronous `read`, and IMO it's nicer that both `read` behave the same
No need to acquire a lock here, because `stopped` is atomic
Because aws async requests cannot be cancelled I tried to check also here. 
The stop flow is a s follows:
1. Stop the S3 client responders (if read has started) - this is done in the d'tor of `S3Stop`
2. Destruct the workers thread pool - this will first set the `stopped` flag to true and wait for the threads to finish

Let's assume that the worker thread is in `async_read`, and the shutdown flow performed the first step but not yet the second. 
Note that the S3 client responders are created only when the S3 client sends async request. Therefore, after the first step new `Responder` objects will not be stopped, and we need to check the `stopped` flag here.